Refactor startup/shutdown events to lifespan events

KhushiRajurkar opened the pull request:
## **Changes Made**
- Replaced deprecated startup/shutdown events with **lifespan events** in `main.py`.  
- Improved **app initialization and cleanup** using FastAPI’s modern event handling.  
- Ensured **environment variable validation** occurs properly during startup.  

## **Issue Reference**
- Closes #152  

## **Notes**
- The application runs **successfully inside the Docker container**.  
- Some **test failures (5 total) appear unrelated** to my changes, mostly concerning authentication and connection issues.  
- This **draft PR is opened for troubleshooting** with maintainers.  

---

## **Changes proposed in this pull request**
✅ Updated FastAPI event handling to **lifespan events**.  
✅ Ensured **vocabulary loading and cleanup** occur within the app lifecycle.  
✅ **Maintains compatibility** with existing API functionality.

## Summary by Sourcery

Migrates the application from deprecated startup/shutdown events to FastAPI lifespan events. This change improves the management of the application lifecycle, including initialization and cleanup tasks such as environment variable validation and vocabulary loading.

Bug Fixes:
- Fixed an issue where environment variable validation was not occurring properly during startup by moving the validation logic to the lifespan event handler.

Enhancements:
- Improved application initialization and cleanup by using FastAPI's lifespan events for managing the application lifecycle.

alyssadai reviewed (changes requested):
Hi @KhushiRajurkar,

Thanks for your PR! The CI tests reproduced 2/5 of the failures you mentioned encountering in the issue: https://github.com/neurobagel/api/actions/runs/13127168535/job/36625878480?pr=407

These are due to your code changes breaking assumptions of our existing unit tests of the app startup checks, which can be fixed by updating either the tests or your code - see my detailed comments below.

As for the other 3 test failures you reported, these are integration tests (marked with `@pytest.mark.integration`, see example [here](https://github.com/neurobagel/api/blob/3718200213d81a07755bcd97365f6a5a2918dd97/tests/test_query.py#L656)), which means that you need a synthetic Neurobagel node running in order to run the tests. The `httpx.Connect` errors you encountered suggest that in the integration tests, requests involving the synthetic node are failing, so the node may not have been set up correctly in your local environment. 

More details about the integration test setup are in our [README](https://github.com/neurobagel/api?tab=readme-ov-file#testing), but actually, I'd recommend simply running the test suite using:
```
pytest
```
which should skip the integration tests by default. If you're developing in a Docker container as opposed to a normal Python environment, you may need extra steps to ensure that the tests can reach the test node (which is its own Docker Compose stack). This may not be worthwhile for the purposes of your PR, which shouldn't affect the integration tests. As a sanity check, you can see that the same integration tests which are failing for you locally are also running and passing in the [CI tests](https://github.com/neurobagel/api/actions/runs/13127168535/job/36625878480?pr=407) on this PR :)

I've left some other comments below, including suggestions for more detail in the new docstrings. Have a look and let me know if you have questions.

KhushiRajurkar commented:
Hello @alyssadai,

Thank you for the detailed review and guidance!
Apologies for these oversights and the delay in responding—I was reviewing the feedback to understand the required changes.
I’ll carefully review the file and make the necessary fixes, including restoring and refactoring the docstrings, correcting the spacing issue in the error message, and ensuring proper formatting with pre-commit. I’ll push the updates soon and re-run the tests to verify everything. 
Thank you!

alyssadai commented:
Sounds great @KhushiRajurkar - please mark your PR as ready to review once you have made the changes!

alyssadai commented:
Hi @KhushiRajurkar, just checking in: do you have any questions about addressing the review comments?

Please feel free to let us know if you no longer have capacity to work on this issue, so that one of the maintainers can take over for your PR.

KhushiRajurkar commented:
> Hi @KhushiRajurkar, just checking in: do you have any questions about addressing the review comments?
> 
> Please feel free to let us know if you no longer have capacity to work on this issue, so that one of the maintainers can take over for your PR.

Hello @alyssadai,
Apologies for the delay in updating. I made the required changes about 5 days ago but encountered issues while running pytest and pre-commit. I wanted to ensure code consistency before pushing the changes.

I believe the issue might be related to my local system environment. As a workaround, I had planned to run the tests in Docker as a final approach before pushing the changes.

I have modified the following files:

`main.py`

`security.py`

`test_security.py`

I’ll proceed with the Docker approach and push the updates soon. Please let me know if there are any alternative suggestions or if you’d like me to take a different approach.

Thank you!

alyssadai commented:
Hi @KhushiRajurkar,

I wouldn't recommend running the tests in Docker, as that will be difficult for us to replicate on our end (our CI tests run just in a bare Python environment).

Instead, I'd suggest first committing and pushing your current changes to this branch (our CI tests will run on them so we can see directly in this PR if there are any tests still failing), and then setting up a fresh Python virtual environment locally to ensure that you have the correct dependencies and linters as defined for this repo.

For example, from the root of this repo:
```bash
python -m venv venv  # ensure you have Python 3.10+
pip install -r requirements.txt

# following https://github.com/neurobagel/api?tab=readme-ov-file#testing
git submodule init
git submodule update

# following https://neurobagel.org/contributing/CONTRIBUTING/#follow-repository-code-style
pre-commit install
pre-commit run
```

and then, you can confirm that the tests run using the command `pytest` (don't worry about the integration tests / setting up Docker Compose).

Let me know if that works.

KhushiRajurkar commented:
> Hi @KhushiRajurkar,
> 
> I wouldn't recommend running the tests in Docker, as that will be difficult for us to replicate on our end (our CI tests run just in a bare Python environment).
> 
> Instead, I'd suggest first committing and pushing your current changes to this branch (our CI tests will run on them so we can see directly in this PR if there are any tests still failing), and then setting up a fresh Python virtual environment locally to ensure that you have the correct dependencies and linters as defined for this repo.
> 
> For example, from the root of this repo:
> 
> ```shell
> python -m venv venv  # ensure you have Python 3.10+
> pip install -r requirements.txt
> 
> # following https://github.com/neurobagel/api?tab=readme-ov-file#testing
> git submodule init
> git submodule update
> 
> # following https://neurobagel.org/contributing/CONTRIBUTING/#follow-repository-code-style
> pre-commit install
> pre-commit run
> ```
> 
> then, you can try running the tests again using the command `pytest` (don't worry about the integration tests / setting up Docker Compose).
> 
> Let me know if that works. For us to better track progress on your PR, I'd encourage you to always push your changes first!

Hello @alyssadai 
Thank you for your guidance! I will commit and push my changes within 24 hours (as I am currently traveling) and will avoid using Docker for testing.

Regarding pytest, I had encountered an issue while running the command: 
```pip install -r requirements.txt```
I was unable to install the dependencies, which prevented me from running pytest in the virtual environment. I'll attempt to resolve this after pushing my changes.

Thank you for your patience and support!

KhushiRajurkar commented:
Hello @alyssadai, I've committed and pushed the changes in PR #407. Please review my changes and provide your valuable feedback.

Thank you!

KhushiRajurkar commented on app/main.py:
```diff
@@ -35,32 +94,21 @@
 
 @app.get("/", response_class=HTMLResponse)
 def root(request: Request):
```
Apologies for the oversight. I'll make the necessary change.

alyssadai reviewed (commented):
Hi @KhushiRajurkar, thanks for the changes! The tests are now passing 🎉 

I left a few final minor suggestions below. I also notice that the pre-commit check is still failing, which I'm assuming may be related to your difficulties installing the dependencies locally.

Since this PR is close to ready, for the sake of being able to merge it soon (we have a couple PRs being worked on in parallel, and we don't want them to get too out of sync and run into merge conflicts), I will apply the remaining changes directly and run pre-commit locally so that we do not merge with the failing check.

> Regarding pytest, I had encountered an issue while running the command:
> `pip install -r requirements.txt`
> I was unable to install the dependencies, which prevented me from running pytest in the virtual environment. I'll attempt to resolve this after pushing my changes.

If you are still having trouble with the dependencies, could you please open an issue or reach out to us on Discord with the exact error message you encountered, so we can better help you debug it? Without seeing the error, it is hard to know whether the issue is related to your Python installation or something else about your dev setup.

alyssadai reviewed (approved):
Ready to merge! 🧑‍🍳 Thanks again @KhushiRajurkar for the contribution! 

As mentioned above. please feel free to open an issue or message us directly on Discord for further help debugging the issue you encountered with installing the requirements.

KhushiRajurkar commented:
> Hi @KhushiRajurkar, thanks for the changes! The tests are now passing 🎉
> 
> I left a few final minor suggestions below. I also notice that the pre-commit check is still failing, which I'm assuming may be related to your difficulties installing the dependencies locally.
> 
> Since this PR is close to ready, for the sake of being able to merge it soon (we have a couple PRs being worked on in parallel, and we don't want them to get too out of sync and run into merge conflicts), I will apply the remaining changes directly and run pre-commit locally so that we do not merge with the failing check.
> 
> > Regarding pytest, I had encountered an issue while running the command:
> > `pip install -r requirements.txt`
> > I was unable to install the dependencies, which prevented me from running pytest in the virtual environment. I'll attempt to resolve this after pushing my changes.
> 
> If you are still having trouble with the dependencies, could you please open an issue or reach out to us on Discord with the exact error message you encountered, so we can better help you debug it? Without seeing the error, it is hard to know whether the issue is related to your Python installation or something else about your dev setup.

Hello @alyssadai,
Thank you so much for your guidance and your patience! I'll take a closer look at the dependency issue and reach out on Discord if I need further assistance.😊
Thank you!

KhushiRajurkar commented:
> Ready to merge! 🧑‍🍳 Thanks again @KhushiRajurkar for the contribution!
> 
> As mentioned above. please feel free to open an issue or message us directly on Discord for further help debugging the issue you encountered with installing the requirements.

Thank you, @alyssadai, for the opportunity to contribute to Neurobagel and for your guidance and support throughout this PR! 🙇🏻‍♀️ This experience has helped me grow as a programmer, and I’m grateful for the chance to learn and improve!
